src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1401,20 +1661,97 @@ class StickyClassDumper : public KlassClosure {
   }
 };
 
+// Large object heap dump support.
+// To avoid memory consumption, when dumping large objects such as huge array and
+// large objects whose size are larger than LARGE_OBJECT_DUMP_THRESHOLD, the scanned
+// partial object/array data will be sent to the backend directly instead of caching
+// the whole object/array in the internal buffer.
+// The HeapDumpLargeObjectList is used to save the large object when dumper scans
+// the heap. The large objects could be added (push) parallelly by multiple dumpers,
+// But they will be removed (popped) serially only by the VM thread.
+class HeapDumpLargeObjectList : public CHeapObj<mtInternal> {
+ private:
+  class HeapDumpLargeObjectListElem : public CHeapObj<mtInternal> {
+   public:
+    HeapDumpLargeObjectListElem(oop obj) : _obj(obj), _next(NULL) { }
+    oop _obj;
+    HeapDumpLargeObjectListElem* _next;
+  };
+
+  HeapDumpLargeObjectListElem* _head;
+  uint _length;
```

Comment:
I don't think the length is needed at all. You just have to know if the list is empty or not and that can be found out via the _head field.